### ISSUE TITLE
BaseSettings Singleton option?

### ISSUE THREAD

**Issue by jasonkuhrt** [samuelcolvin/pydantic#586]
I've noticed a pattern come up in several of our projects and thought maybe (but I'm really not sure) pydantic could ship support for it.

```py
# config.py
from functools import lru_cache
from pydantic import BaseSettings


class Config(BaseSettings):
    a: str
    b: str
    c: str

@lru_cache()
def read() -> Config:
    return Config()
```
```py
# a.py
from . import config
conf = config.read
```
```py
# b.py
from . import config
conf = config.read
```

Would it be a good library addition if we could do something like:
```py
# config.py
from functools import lru_cache
from pydantic import BaseSettings


class Config(BaseSettings):
    a: str
    b: str
    c: str

    class Config:
        singleton = True
```
```py
# a.py
from . import Config
conf = Config()
```
```py
# b.py
from . import config
conf = Config()
```

I definitely appreciate that pydantic should not try to take care of application-level features like dependency injection and object caching. However, for the special case of `BaseSettings` this just seems like a practical feature. Thoughts?

Maybe there is already a more general solution here that is readily available that i missed, allowing e.g.:

```py
# config.py
from pydantic import BaseSettings

@singleton # exists?
class Config(BaseSettings):
    a: str
    b: str
    c: str
```

But I haven't come across that yet. And if such a thing is from another third party lib. I still feel then that there is value in pydantic shipping an out of box solution, if its what users would want most of the time.

**Comment by dmontagu** [comment]
I have personally dealt with this by just using a module-level instance.

My settings file looks like this:
```python
# config.py
from pydantic import BaseSettings

class AppSettings(BaseSettings):
    ...

settings = AppSettings()
```
Then, when I want to access the settings object, I just do
```python
from .config import settings
```
and it imports that specific instance.

As far as I can tell, the only difference between this and your current approach is that the module-variable approach instantiates the config on import, whereas your approach waits to instantiate it until the `config.read()` function is called for the first time. In my applications, 1) the config is used very early in the app lifecycle, and 2) I never expect the environment to change anyway, so this difference is irrelevant. But I could imagine an application requiring more of a lazy-loading approach.

Would this approach be undesirable for you?

**Comment by samuelcolvin** [comment]
I completely agree `settings` should be a singleton, however I've never had any problem creating it as a singleton which is then used throughout the application.

This is probably because most applications I build are either based on aiohttp or use aiohttp's approach of a single "app" context. 

See for example [here](https://github.com/samuelcolvin/nosht/blob/master/py/web/main.py#L71). This approach also works well for testing where you generally want a very different settings instance which can easily be changed on a per-test basis without reloading modules.

My suspicion is that you've made design mistakes if you find yourself trying to initialise settings more than once. Probably best you fix the root mistake or take care of the workaround outside pydantic.

**Comment by jasonkuhrt** [comment]
Makes sense and thanks!

**Comment by jasonkuhrt** [comment]
@dmontagu that doesn't work for my because it would lead to the config being calculated before environment mocking takes place by pytest.

@samuelcolvin agree; There is a discussion on the [fastapi project about similar](https://github.com/tiangolo/fastapi/issues/190) but as far as I know still unresolved.

**Comment by dmontagu** [comment]
@jasonkuhrt makes sense, thanks for explaining

**Comment by tl24** [comment]
I usually put these in a factory function with lru_cache

```
from functools import lru_cache

@lru_cache(maxsize=1)
def get_settings():
   return AppSettings()
```

**Comment by winstxnhdw** [comment]
Here's a better and type-safe way to do it.

```python
def singleton(cls: type[T]) -> T:
    """
    Summary
    -------
    a decorator to make a class a singleton
    """
    return cls()

@singleton
class Config(BaseSettings):
    secret: str

print(Config.secret)
```

**Comment by MarcBresson** [comment]
Hey, sorry to dig up this issue.

I feel like having singletons can lead to more elegant code such as

```python
# config.py
from pydantic import BaseSettings

class AppSettings(BaseSettings):
    hostname: str
    port: int
    secure: bool
```

```python
# michel.py
from src.config import AppSettings

OneSuperSecretService(
    hostname=AppSettings().hostname,
    port=AppSettings().port,
    secure=AppSettings().secure,
)
```

versus the less elegant version

```python
# config.py
from pydantic import BaseSettings

class AppSettings(BaseSettings):
    hostname: str
    port: int
    secure: bool

app_settings = AppSettings()
```

```python
# michel.py
from src.config import app_settings

OneSuperSecretService(
    hostname=app_settings.hostname,
    port=app_settingsport,
    secure=app_settings.secure,
)
```

**Comment by nathan-chappell** [comment]
@samuelcolvin I guess for me the question is sort of *when does the settings get initialized?*  It's typical that you just want to create one and leave it in the module where it is defined.  But perhaps you want to import the file where the settings class is located, but you don't necessarily want the instance created right away.

I'm not trying to reopen the issue, `pydantic` saying "leave the problem in userspace" is pretty reasonable to me, I just wanted to try to argue the case that it's not necessarily an indication of a design flaw, and offer a simple solution I'll use:

```py
TSettings = TypeVar("TSettings", bound=BaseSettings)

class SingletonBaseSettings(BaseSettings):
    @classmethod
    @cache
    def instance(cls: type[TSettings], **kwargs) -> TSettings:
        return cls(**kwargs)
```

One side thing that's kinda nice is my IDE doesn't complain about `.instance()` like it complains about `SomeSettingsClass()`.  Usually I have to tell the IDE to shut-up about this:

```
class Settings(SingletonBaseSettings):
    FOO: str

print(Settings())           # type: ignore...  If I'm missing something I want an error raised!
print(Settings.instance())  # Ahh, no complaints (and no safety!)
```

Looking at it, it's probably a little *too* opinionated, but oh well.

**Comment by nathan-chappell** [comment]
@MarcBresson It's probably debatable about which code is more elegant, but to get the same effect I've defined my own base class as follows:

```py
T = TypeVar("T")
TSettings = TypeVar("TSettings", bound=BaseSettings)


def _cache(t: T) -> T:
    return cache(t)  # type: ignore


class SingletonBaseSettings(BaseSettings):
    @classmethod
    @_cache
    def instance(cls: type[TSettings], **kwargs) -> TSettings:
        return cls(**kwargs)

    i = instance
```

Used like so:

```py
class Settings(SingletonBaseSettings):
    FOO: str

    # This model_post_init and instance_count is for demonstration only
    instance_count: ClassVar[int] = 0

    def model_post_init(self, __context: Any) -> None:
        Settings.instance_count += 1
        return super().model_post_init(__context)

# first instance
print(Settings())  # type: ignore...  If I'm missing something I want an error raised!

# second instance
print(Settings.i())
# third instance
print(Settings.i(FOO="bar"))
print(Settings.i().FOO, Settings.instance_count)
print(Settings.i(FOO="bar").FOO, Settings.instance_count)

# FOO='env_foo'
# FOO='env_foo'
# FOO='bar'
# env_foo 3
# bar 3
```

**Comment by MarcBresson** [comment]
@nathan-chappell thank you for your reply!

Is there any reason why you don't overload the __new__ method in your `Settings` class to enable singleton throughout the usual class interface? I have not been successful in doing that, so it may not be possible.

**Comment by nathan-chappell** [comment]
@MarcBresson The *official* reason is that this way it is "opt-in."  The real reason is that I know `Pydantic` does some meta-class funny business, and I didn't want to spend too much time to worry about how it interacts with trying to override `__new__`.  

There seems to be much disagreement about the best way to do it: [Is there a simple, elegant way to define singletons?](https://stackoverflow.com/questions/31875/is-there-a-simple-elegant-way-to-define-singletons), and any way seems to lead to a surprising edge case.

This seems to work:

```py
class SingletonBaseSettings2(BaseSettings):
    _instance: ClassVar[Any | None] = None

    def __new__(cls: type[Self], *args, **kwargs) -> Self:
        if cls._instance is None:
            cls._instance = super().__new__(cls, *args, **kwargs)
        return cls._instance
```

 but I wouldn't be surprised if there are surprising implications of doing it this way.

**Comment by MarcBresson** [comment]
Yeah, indeed initialization always happens after `__new__`, but it is possible to add the following to your singleton (and all its inherited classes):

```python
...
    def __init__(self, *args, **kwargs):
        if self.instance is not None:
            return

        super().__init__(*args, **kwargs)
```

either way, metaclasses indeed ruin the thing. I tried implementing a singleton BaseSetting and it did not work (I compared it to a non-singleton BaseSettings that did work).


```python
from typing import ClassVar, Any

class SingletonBaseSettings2(BaseSettings):
    _instance: ClassVar[Any | None] = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super().__new__(cls, *args, **kwargs)
        return cls._instance
    
    def __init__(self, *args, **kwargs):
        if self._instance is not None:
            return
    
        super().__init__(*args, **kwargs)


class FooBar(SingletonBaseSettings2):
    version: str

    def __init__(self):
        if self._instance is not None:
            return
        print("hello from FooBar")

class Test(BaseSettings):
    version: str
```

```python
print(FooBar() is FooBar())
>> "hello from FooBar"
>> True
```

```python
print(FooBar().version)
>> AttributeError: 'FooBar' object has no attribute 'version'
print(Test().version)
>> 0.5.56
```

**Comment by winstxnhdw** [comment]
FWIW here's a more elegant and performant implementation I came up with a long time ago:

```python
# config.py
from pydantic_settings import BaseSettings


def singleton[T](cls: type[T]) -> T:
    return cls()


@singleton
class Config(BaseSettings):
    server_port: int = 49494
    server_root_path: str = '/api'
    worker_count: int = 1
    use_cuda: bool = False
```

Then you can use this anywhere like so:

```python
from config import Config

print(Config.worker_count) # 1
```

**Comment by winstxnhdw** [comment]
Oh, you're right. It is calling `Config()` each time. I guess we do have to cache it in the end 😢

**Comment by nathan-chappell** [comment]
@MarcBresson I think requiring the user to follow some protocol like that is another pretty good reason not to implement it that way.  I was originally thinking of responding with the following idea, but then it doesn't work with inheritance:

```py
@cache
def _build_object(t: type[type], *args, **kwargs) -> Any:
    return object.__new__(t)


def singleton_decorator(t: type[TSettings]) -> type[TSettings]:
    instance_ids: set[int] = set()
    t_new = t.__new__
    t_init = t.__init__

    @wraps(t_new)
    def __new_new__(*args, **kwargs) -> TSettings:
        if t_new is object.__new__:
            instance = _build_object(t, *args, **kwargs)
        else:
            instance = t_new(t, *args, **kwargs)

        instance_id = id(instance)
        if instance_id not in instance_ids:
            print("new miss")
            t_init(instance, *args, **kwargs)  # type: ignore
            instance_ids.add(instance_id)
        return instance

    @wraps(t_init)
    def __new_init__(self: TSettings, *args, **kwargs) -> None:
        if id(self) in instance_ids:
            return
        raise Exception("UNREACHABLE")

    t.__new__ = __new_new__
    t.__init__ = __new_init__  # type: ignore
    return t
```

Example:

```py
@singleton_decorator
class Settings2(BaseSettings):
    FOO: str


assert Settings2() is Settings2()
assert Settings2() is not Settings2(FOO="custom-foo")
assert Settings2(FOO="custom-foo") is Settings2(FOO="custom-foo")
```

**Comment by winstxnhdw** [comment]
If you're okay with caching,

```python
# config.py
from functools import lru_cache

from pydantic_settings import BaseSettings

@lru_cache(maxsize=1)
def singleton[T](cls: type[T]) -> T:
    return cls()


@singleton
class Config(BaseSettings):
    server_port: int = 49494
    server_root_path: str = '/api'
    worker_count: int = 1
    use_cuda: bool = False
    
assert Config.worker_count is Config.worker_count
```

**Comment by MarcBresson** [comment]
I thought I understood how `lru_cache(maxsize=1)` was working, but apparently not. I tried creating multiple class using the `@singleton` decorator to overload the cache, but I could not generate an error.

```python
from functools import lru_cache

from pydantic_settings import BaseSettings

@lru_cache(maxsize=1)
def singleton[T](cls: type[T]) -> T:
    return cls()
```

```python
@singleton
class Config1(BaseSettings):
    logname: str = ""

@singleton
class Config2(BaseSettings):
    logname: str = ""

@singleton
class Config3(BaseSettings):
    logname: str | None = None
```

```python
logname1 = Config1.logname
logname2 = Config2.logname
logname3 = Config3.logname

c1 = Config1
c2 = Config2
c3 = Config3

assert logname1 is Config1.logname
assert logname3 is Config3.logname
assert logname2 is Config2.logname

assert Config1 is c1
assert Config2 is c2
assert c3 is Config3
# everything passes :(
```

**Comment by nathan-chappell** [comment]
@winstxnhdw There's no principle difference from this:

```py
class Config(BaseSettings):
    server_port: int = 49494
    server_root_path: str = '/api'
    worker_count: int = 1
    use_cuda: bool = False

Config = Config()
```

One of the issues I was trying to deal with was controlling *when* the instance gets created (in particular, not when the module containing the class is imported).  I'd also like the class to pretty much behave like a class, but only ever make one instance (I'm including instances with different initialization parameters in my code, but if you are willing to never pass anything to init then it's not necessary).

Basically, how do you implement `singleton` with the *least amount of surprise*

**Comment by winstxnhdw** [comment]
> @winstxnhdw It is extremely short indeed. It's a bit weird that the interface is with the class itself.
> 
> Are you certain this is a proper singleton? I feel like Config is getting recreated and initialized every single time we access one of its attribute.
> 
> Maybe you could use a cache to create a false singleton.

I circled back to this the other day and I found that it does not reinitialise on each access. Decorators only run once on start-up.

**Comment by MarcBresson** [comment]
@samuelcolvin would you mind re-evaluating this feature request? Workarounds to create singletons are complex because of the very low level of pydantic (with metaclasses, and Cython, etc.).

> The best option would be to have the singleton behaviour designed from within pydantic_settings (maybe from the metaclasses themselves). We could think of something like that for the end user:
> 
> ```python
> from pydantic_settings import BaseSettings, SettingsConfig
> 
> class MySingletonConfig(BaseSettings):
>     model_config = SettingsConfig(singleton=True)
>     ...
> ```

It would also have the benefits of being initialized on the first call, and not when the module is being imported the first time.

**Comment by climber-dialectical** [comment]
> [@MarcBresson](https://github.com/MarcBresson) I think requiring the user to follow some protocol like that is another pretty good reason not to implement it that way. I was originally thinking of responding with the following idea, but then it doesn't work with inheritance:
> 
> @cache
> def _build_object(t: type[type], *args, **kwargs) -> Any:
>     return object.__new__(t)
> 
> 
> def singleton_decorator(t: type[TSettings]) -> type[TSettings]:
>     instance_ids: set[int] = set()
>     t_new = t.__new__
>     t_init = t.__init__
> 
>     @wraps(t_new)
>     def __new_new__(*args, **kwargs) -> TSettings:
>         if t_new is object.__new__:
>             instance = _build_object(t, *args, **kwargs)
>         else:
>             instance = t_new(t, *args, **kwargs)
> 
>         instance_id = id(instance)
>         if instance_id not in instance_ids:
>             print("new miss")
>             t_init(instance, *args, **kwargs)  # type: ignore
>             instance_ids.add(instance_id)
>         return instance
> 
>     @wraps(t_init)
>     def __new_init__(self: TSettings, *args, **kwargs) -> None:
>         if id(self) in instance_ids:
>             return
>         raise Exception("UNREACHABLE")
> 
>     t.__new__ = __new_new__
>     t.__init__ = __new_init__  # type: ignore
>     return t
> Example:
> 
> @singleton_decorator
> class Settings2(BaseSettings):
>     FOO: str
> 
> 
> assert Settings2() is Settings2()
> assert Settings2() is not Settings2(FOO="custom-foo")
> assert Settings2(FOO="custom-foo") is Settings2(FOO="custom-foo")





https://pythondev.space/questions/solitude-apex-class-and-issue-with-creating-an-instance

**Comment by Docdez** [comment]
> Yeah, indeed initialization always happens after `__new__`, but it is possible to add the following to your singleton (and all its inherited classes):


Seems additional class attribute is needed to make this workable.

```python

import os

from typing import ClassVar, Optional
from pydantic_settings import BaseSettings, SettingsConfigDict

class BaseSettingsSingleton(BaseSettings):
    _sn_cls: ClassVar[Optional["BaseSettingsSingleton"]] = None
    _sn_is_init: ClassVar[bool] = False

    def __new__(cls, *args, **kwargs):
        if cls._sn_cls is None:
            print("DEBUG: NEW call")
            cls._sn_cls = super().__new__(cls, *args, **kwargs)
        return cls._sn_cls

    def __init__(self, *args, **kwargs):
        if self._sn_is_init:
            return
        super().__init__(*args, **kwargs)
        print("DEBUG: INIT call")
        self.__class__._sn_is_init = True

    @classmethod
    def clear_singleton(cls) -> None:
        cls._sn_cls = None
        cls._sn_is_init = False


class TestOne(BaseSettingsSingleton):
    VAR: str

    model_config = SettingsConfigDict(env_prefix="TEST_1_")


class TestTwo(BaseSettingsSingleton):
    VAR: str

    model_config = SettingsConfigDict(env_prefix="TEST_2_")

```

```python
>>> one = TestOne()
DEBUG: NEW call
VAR
  Field required [type=missing, input_value={}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.10/v/missing

>>> os.environ["TEST_1_VAR"] = "test_one"
>>> one = TestOne()
DEBUG: INIT call

>>> one is TestOne()
True

>>> one.VAR
'test_one'

>>> os.environ["TEST_2_VAR"] = "test_two"
>>> two = TestTwo()
DEBUG: NEW call
DEBUG: INIT call
```

**Comment by GrayWizard12345** [comment]
What about just defining module level singletone with lazy init?
```py
class MySettings(BaseSettings):
    name: str = "postgres"
    user: str = "postgres"
    password: str = "postgres"
    host: str = "localhost"
    port: int = 5432

    class Config:
        env_prefix = "DB_"

...
...

_settings: MySettings | None = None
def Settings() -> MySettings:
    global _settings
    if not _settings:
        _settings = MySettings()
    return _settings
```

And avoid creating global vars that reference this function in modules that require testing, so dont do
```py
from settings import Settings
settings_instance = Settings()
```
And better call Settings() in functions/methods instead. This will help avoid caveat with tests
